pkg/util/hostname.go

```diff
@@ -101,6 +102,37 @@ func setHostnameProvider(name string) {
 	inventories.SetAgentMetadata("hostname_source", name)
 }
 
+// isOSHostnameUsable returns `false` if it has the certainty that the agent is running
+// in a non-root UTS namespace because in that case, the OS hostname characterizes the
+// identity of the agent container and not the one of the nodes it is running on.
+// There can be some cases where the agent is running in a non-root UTS namespace that are
+// not detected by this function (systemd-nspawn containers, manual `unshare -u`…)
+// In those uncertain cases, it returns `true`.
+func isOSHostnameUsable() (osHostnameUsable bool) {
+	// If the agent is not containerized, just skip all this detection logic
+	if !config.IsContainerized() {
+		return true
+	}
+
+	// Check UTS namespace from docker
+	utsMode, err := GetAgentUTSMode()
+	if err == nil && (utsMode != containers.HostUTSMode && utsMode != containers.UnknownUTSMode) {
+		log.Debug("Agent is running in a docker container without host UTS mode: OS-provided hostnames cannot be used for hostname resolution.")
+		return false
+	}
+
+	// Check hostNetwork from kubernetes
+	// because kubernetes sets UTS namespace to host if and only if hostNetwork = true:
+	// https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/dockershim/security_context.go#L203-L205
```

Comment:
```suggestion
	// https://github.com/kubernetes/kubernetes/blob/cf16e4988f58a5b816385898271e70c3346b9651/pkg/kubelet/dockershim/security_context.go#L203-L205
```

Reply:
pinning to a commit instead of master to avoid pointing to something irrelevant as soon as this file gets updated upstream